### PR TITLE
Snova chenw/enhance chat template

### PR DESCRIPTION
**Pull request by snova-chenw**
## Summary
Allow users to use apply chat template with our input data format [Docs](https://huggingface.co/docs/transformers/chat_templating)

We can make a follow up PR that also allows user to pass in data in "user" "assistant" format. 

## PR Checklist
- [x] My PR is less than 500 lines of code
- [x] I have added sufficient comment as docstrings in my code
- [x] I have made corresponding changes to the documentation
- [x] I have added examples for test

### CODE REVIEW

File: generative_data_prep/processors/article_tokenizer.py [hunk 1]
```diff
@@ -141,6 +147,193 @@ def _update_token_metrics(self, tokenized_sequences: List[TokenizedSequence]):
...
+        despite the transformation of the [prompt, completion] pair into a raw string format with additional instruction tokens.
+
+        The following steps are performed:
+        1. Recognizes and preserves the original dialogue data format. (<prompt_placeholder>, <completion_placeholder>)
+        2. Converts the original [prompt, completion] format to the standard conversation [role, content] format.
+            {"prompt": "Hello, how are you?", "completion": "I'm doing great"} ->
+                {"role": "user", "content": "Hello, how are you?"}, {"role": "assistant", "content": "I'm doing great."}
+        """
+        for prompt_completion in jsonl:
+            if self.prompt_keyword not in prompt_completion:
```
Comment (snova-zoltanc) [inline generative_data_prep/processors/article_tokenizer.py:194]: We actually allow empty prompts in the other version of process_jsonl, sometimes users only want to pretrain or mix pretraining data without any prompts. Lets allow them to not provide a prompt, and default to empty string? 

prompt = prompt_completion[self.prompt_keyword] if self.prompt_keyword in prompt_completion else ""

File: generative_data_prep/processors/article_tokenizer.py [hunk 2]
```diff
@@ -141,6 +147,193 @@ def _update_token_metrics(self, tokenized_sequences: List[TokenizedSequence]):
             self.metrics.completion_tokens += seq.completion_tokens
             self.metrics.padding_tokens += seq.pad_tokens
 
+    def process_jsonl_with_chat_template(self, jsonl: List[dict]) -> List[TokenizedArticle]:
```
Comment (snova-zoltanc) [inline generative_data_prep/processors/article_tokenizer.py:150]: We need to add pytests with unit tests to test this, similar to 

tests/test_article_tokenizer.py

**Review by snova-zoltanc — CHANGES_REQUESTED** [review]
Great work this looks awesome, I left initial review and I am happy to help with some of these changes. So far I have just requested code quality updates, tests and cleanup. 

I will finish stepping through the new code in the debugger in a future PR

File: generative_data_prep/processors/article_tokenizer.py [hunk 3]
```diff
@@ -141,6 +148,200 @@ def _update_token_metrics(self, tokenized_sequences: List[TokenizedSequence]):
...
+                err_msg = f"Prompt keyword required in every jsonl, {self.prompt_keyword} not found"
+                raise json.JSONDecodeError(err_msg, str(jsonl), 0)
+
+            if self.completion_keyword not in prompt_completion:
+                err_msg = f"Completion keyword required in every jsonl, {self.completion_keyword} not found"
+                raise json.JSONDecodeError(err_msg, str(jsonl), 0)
+
+            prompt = prompt_completion[self.prompt_keyword]
+            completion = prompt_completion[self.completion_keyword]
+            converted_jsonl.append({"role": "user", "content": DEFAULT_PROMPT_PLACEHOLDER})
```
Comment (snova-bol) [inline generative_data_prep/processors/article_tokenizer.py:209]: Do we want to hard code "user"/"assistant" here or do we want to allow users to define these? Also, do we want to consider the case when user wants to pass in system prompt?

File: generative_data_prep/processors/article_tokenizer.py [hunk 4]
```diff
@@ -141,6 +148,200 @@ def _update_token_metrics(self, tokenized_sequences: List[TokenizedSequence]):
...
+
+                new_completion = completion
+                while split_chat and (
+                    split_chat[0] != DEFAULT_PROMPT_PLACEHOLDER
+                    or (start_instr_of_chat and split_chat[0] != start_instr_of_chat)
+                ):
+                    new_completion += split_chat.pop(0)
+
+                # remove the duplicated eos tokens from new_completion as we will add it later
+                new_completion = new_completion.replace(self.tokenizer.eos_token, "")
```
Comment (snova-bol) [inline generative_data_prep/processors/article_tokenizer.py:318]: Will there be a corner case if eos is in the middle of the completion?

File: generative_data_prep/processors/article_tokenizer.py [hunk 5]
```diff
@@ -141,6 +147,193 @@ def _update_token_metrics(self, tokenized_sequences: List[TokenizedSequence]):
...
+                # Should not reach here
+                raise ValueError(f"Invalid token {token}")
+                # Take care of the case where the chat template introduces additional tokens.
+                # tokens += self.tokenize(completion=token, prompt=None, category_id=category_id)
+
+        if len(tokens) > 0:
+            tokens[-1].make_article_boundary()
+        tokenized_articles.append(TokenizedArticle(tokens))
+
+        return tokenized_articles
```
[the remark this answers is not in the document]
Comment (snova-bol) [reply generative_data_prep/processors/article_tokenizer.py:343]: +1 on this

File: generative_data_prep/utils/utils.py [hunk 6]
```diff
@@ -244,6 +244,13 @@ def data_prep_arg_builder(parser: argparse.ArgumentParser):
             " file that contains a list of all the strings of the 'category' keys in your dataset."
         ),
     )
+    parser.add_argument(
+        "--apply_chat_template",
+        action="store_true",
+        help="If you want to apply the chat template to your data, include this flag. \
+        The chat template is a template that is applied to the data to make it more conversational \
+        which includes a role, like “user” or “assistant”, as well as message text.",  # noqa: E251
```
[the remark this answers is not in the document]
Comment (snova-bol) [reply generative_data_prep/utils/utils.py:252]: +1

File: generative_data_prep/processors/article_tokenizer.py [hunk 7]
```diff
@@ -141,6 +148,200 @@ def _update_token_metrics(self, tokenized_sequences: List[TokenizedSequence]):
...
+            err_msg = f"Number of prompts and completions must be equal, \
+             {len(prompts)} prompts and {len(completions)} completions found"
+            raise ValueError(err_msg)
+
+        if len(prompts) != len(jsonl):
+            err_msg = f"Number of prompts and jsonl must be equal, {len(prompts)} prompts and {len(jsonl)} jsonl found"
+            raise ValueError(err_msg)
+
+        # Apply specified chat templates using tokenizer (apply_chat_template) without tokenization.
+        formatted_chat = self.tokenizer.apply_chat_template(converted_jsonl, tokenize=False)
```
Comment (snova-bol) [inline generative_data_prep/processors/article_tokenizer.py:225]: Let's add an assertion here to make sure that tokenizer does have apply_chat_template defined?

File: generative_data_prep/processors/article_tokenizer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (snova-chenw) [reply generative_data_prep/processors/article_tokenizer.py]: Thanks for pointing this out, I have revised the code with the latest commit.

File: generative_data_prep/processors/article_tokenizer.py [hunk 8]
```diff
@@ -141,6 +147,193 @@ def _update_token_metrics(self, tokenized_sequences: List[TokenizedSequence]):
...
+            jsonl = [jsonl]
+
+        converted_jsonl = []
+        prompts = []
+        completions = []
+
+        # TODO: Add a trick here to get the generation_prompt in tokenizer
+        # This is a temporary solution, and the tokenizer should be modified in the future.
+        # The tokenizer should be able to get the generation_prompt from the input text.
+        # self.tokenizer.chat_template = "{% for message in messages %} {{'<|im_start|>' + message['role'] + '\n' + message['content'] + '<|im_end|>' + '\n'}}{% endfor %}"
```
[the remark this answers is not in the document]
Comment (snova-chenw) [reply generative_data_prep/processors/article_tokenizer.py]: 'generation_prompt' is from here
https://huggingface.co/docs/transformers/chat_templating#what-are-generation-prompts

File: tests/examples/dialogue/data_prepped_dialogue_chat_template.hdf5
[the remark this answers is not in the document]
Comment (snova-chenw) [reply tests/examples/dialogue/data_prepped_dialogue_chat_template.hdf5:1]: agree, thanks for the help, please feel free to add a test

File: tests/examples/dialogue/decoded_data_prepped_dialogue_chat_template.txt [hunk 9]
```diff
@@ -0,0 +1,682 @@
+================================================================================
```
[the remark this answers is not in the document]
Comment (snova-chenw) [reply tests/examples/dialogue/decoded_data_prepped_dialogue_chat_template.txt:1]: the chat template is applied, and the default chat template in our tokenizer is "{% for message in messages %}{{ message.content }}{{ eos_token }}{% endfor %}"

The tiny difference from the original (without chat_template) is that there is an eos_token at the end of the prompt.

File: README.md [hunk 10]
```diff
@@ -334,6 +334,13 @@ python3 -m generative_data_prep pipeline --input_file_path=./tests/examples/dial
 
 > [View decoded output](tests/examples/dialogue/decoded_data_prepped_dialogue.txt)
 
+#### Example command with applying chat template
```
[the remark this answers is not in the document]
Comment (snova-chenw) [reply README.md:337]: Agree, thanks Zoli, please feel free to update it.

File: generative_data_prep/processors/article_tokenizer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (snova-chenw) [reply generative_data_prep/processors/article_tokenizer.py:343]: Agree! please feel free to refactor this function.

File: generative_data_prep/processors/article_tokenizer.py (hunk 1 above)
Comment (snova-chenw) [reply generative_data_prep/processors/article_tokenizer.py:199]: agree, I have revised the code with the latest commit.

File: generative_data_prep/processors/article_tokenizer.py (hunk 2 above)
Comment (snova-chenw) [reply generative_data_prep/processors/article_tokenizer.py]: Agree, please help to add the unit tests, thanks a lot

File: generative_data_prep/processors/article_tokenizer.py (hunk 3 above)
Comment (snova-chenw) [reply generative_data_prep/processors/article_tokenizer.py:209]: thanks for pointing this out, the format is hard coded here only according to hugging face version.

File: generative_data_prep/processors/article_tokenizer.py (hunk 4 above)
Comment (snova-chenw) [reply generative_data_prep/processors/article_tokenizer.py:318]: thanks, good point, I am going to fix it in the latest commit

File: generative_data_prep/processors/article_tokenizer.py (hunk 7 above)
Comment (snova-chenw) [reply generative_data_prep/processors/article_tokenizer.py:229]: Thanks for pointing this out, I have revised the code in the latest commit.

**Comment by snova-zoltanc** [comment]
The remaining items to complete are

1. Clean up code to break up functions even more
2. Finish pushing new e2e tests
3. Create unit tests for new functions
4. Update the README with how to use apply chat template

**Review by overwindows — APPROVED** [review]

File: README.md [hunk 11]
```diff
@@ -144,6 +150,11 @@ We also support lists of prompt/completion pairs within a `.jsonl` file. This gu
 
 If the JSON objects in your `.jsonl` contain keywords other than **prompt** and **completion**, refer to the `prompt_keyword` and `completion_keyword` flags [below](#prompt_keyword)
 
+### Preparing data for Chat or Instruct Models
+Many chat and instruct models require very specific formatting to input multi turn conversations for training and inference using the [tokenizer.apply_chat_template function](https://huggingface.co/docs/transformers/main/en/chat_templating). We support tokenizing your data in this format, to do so please prepare your data in jsonl format as specified above, and then inlcude the `--apply_chat_template` flag to automatically prepare your data in this format.
```
Comment (snova-connorm) [inline README.md:154]:
```suggestion
Many chat and instruct models require very specific formatting to input multi turn conversations for training and inference. The [tokenizer.apply_chat_template function](https://huggingface.co/docs/transformers/main/en/chat_templating) easily adapts your jsonl data to this format. To use this feature, prepare your data in jsonl format as specified above, and then include the `--apply_chat_template` flag to automatically prepare your data in this format.
```

Light wordsmithing - the first sentence makes it seem like using the `apply_chat_template` requires a specific formatting, when in reality - IIUC - it's the chat/instruction tuned models that require this format and this feature helps prepare your data for that.

File: README.md [hunk 12]
```diff
@@ -266,6 +277,7 @@ This section outlines all the flags you can set to customize the data prep pipel
 | `attention_boundary` | str | 'jsonl' | ['jsonl', 'prompt_completion_pair'] | The boundary to use when training with --article_attention flag. If you choose prompt_completion_pair tokens will only attend to tokens in the prompt_completion_pair. If you choose jsonl, then tokens will attend to all the prompt completion pairs in the jsonl |
 | `prompt_keyword` <span id="prompt_keyword"></span> | str | 'prompt' | | If your input json has a string keyword for prompt other than "prompt", place the keyword here. e.g Input_json: {"source": ... "target": ...} ->`prompt_keyword`='source'. |
 | `completion_keyword` | str | 'completion' | | If your input json has a string keyword for completion other than "completion", place the  keyword here. e.g Input_json: {"source": ... "target": ...} -> --completion_keyword='target'. |
+| `apply_chat_template` | bool | False | | Whether to tokenize the data using tokenizer.apply_chat_template, to add the chatML tags. The tokenizer must have apply_chat_template defined. |
```
Comment (snova-connorm) [inline README.md:280]: Few nits here:
1. Would be nice to add a little snippet on what `chatML` tags are here. I'm not sure users will all know what this means and I think it's too much info to throw into the smaller table above.
2. Not exactly sure what is meant by:
> The tokenizer must have apply_chat_template defined.

Does every new tokenizer need to implement the `apply_chat_template` base class or is there a supported list of tokenizers? A little more elaboration here would be super helpful.

3. Small formatting nit:

```suggestion
| `apply_chat_template` | bool | False | | Whether to tokenize the data using `tokenizer.apply_chat_template`, to add the chatML tags. The tokenizer must have apply_chat_template defined. |
```

**Review by snova-connorm — APPROVED** [review]
Looks good to me. Great addition here, just a few nits for the readme!

**Review by snova-zoltanc — APPROVED** [review]
LGTM

**Review by snova-zoltanc — APPROVED** [review]
LGTM

**Pull request merged**